signatures: Allow additional key version-string characters.

jevolk opened the pull request:
First rust :v:

Feel free to change it or kick it back until it's right.

jevolk committed:
signatures: Relax key version string character constraints.

ShadowJonathan reviewed (changes requested):
While I don't think this is much a problem, i'd still like to see a source on where these things are allowed by the matrix spec (or transitively by another spec, referenced by the matrix spec).

If you can add those as a comment above the additional characters, that'll help.

---

Taking a quick look at the matrix spec, it is silent on this (as a far as i can see), so technically we'd have to accept any character here.

I think that'd be a great case for `compat`, to - under that crate feature - change the logic to accept any non-`:` character for parsing, making it an opaque string.

jplatte commented:
@ShadowJonathan This was recently discussed in #matrix-spec. I don't know about the extra chars for sure, but there is a section in server-server that lists the characters we currently support, and Travis said this should be copied to appendices. I'm not aware of anybody actually putting in the time to update the spec since that discussion though.

ShadowJonathan commented:
I'm also finding https://spec.matrix.org/v1.6/server-server-api/#get_matrixkeyv2server, but like jonas said, there's [some additional discussion](https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$rxNSy1w8HFJMoROofxhOv7ltj7tr5PkRsqdt9RAO_IE?via=matrix.org&via=libera.chat&via=element.io) in #spec after that point.

ShadowJonathan commented on crates/ruma-signatures/src/lib.rs:
```diff
@@ -85,7 +85,9 @@ fn split_id(id: &str) -> Result<(Algorithm, String), Error> {
+
+    if !version.bytes().all(|ch| ch.is_ascii_alphanumeric() || VERSION_ALLOWED.contains(&ch)) {
         return Err(Error::InvalidVersion(version.into()));
     }
```
```suggestion
    #[cfg(feature = "compat")]
    const EXTRA_ALLOWED: [u8; 3] = [b'_', b'+', b'/'];
    #[cfg(not(feature = "compat"))]
    const EXTRA_ALLOWED: [u8; 1] = [b'_'];
      
    if !version.bytes().all(|ch| ch.is_ascii_alphanumeric() || EXTRA_ALLOWED.contains(&ch)) {
        return Err(Error::InvalidVersion(version.into()));
    }
```

ShadowJonathan reviewed (dismissed):
Given the fact that these characters exist in the wild already, but that the spec *does* hint at only allowing `_` as extra character (and no more), I'm going the conservative route and suggesting this be added only under `compat` (which conduit compiles with), to make sure that we draw a line between "what the spec says" and "what (popular) reality is like"

I'm sure that the spec will be updated to reflect this, but in the meantime I don't think this is as 'official' a behaviour, and i'd like to wait and see what is indeed written down after travis has done the relevant research, and not try to predict what the spec will do next.

jevolk committed:
Add extra allowed characters in `compat` mode.

This complies with the published specification by default, with optional compatibility for additional federation keys in use.

Co-authored-by: Jonathan de Jong <jonathandejong02@gmail.com>

jplatte reviewed (approved):
Thanks!

jplatte commented on crates/ruma-signatures/src/lib.rs:
```diff
@@ -85,7 +85,12 @@ fn split_id(id: &str) -> Result<(Algorithm, String), Error> {
+    const EXTRA_ALLOWED: [u8; 3] = [b'_', b'+', b'/'];
+    #[cfg(not(feature = "compat"))]
+    const EXTRA_ALLOWED: [u8; 1] = [b'_'];
+      
```
```suggestion

```

jplatte committed:
Remove trailing whitespace

jplatte committed:
Add compat feature to ruma-signatures
